In [1]:
from lib.project_5_ADH import load_data_from_database, make_data_dict, general_model, general_transformer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

# Step 1 - Benchmarking

**NOTE: EACH OF THESE SHOULD BE WRITTEN SOLELY WITH REGARD TO STEP 1 - BENCHMARKING**

### Domain and Data

The data, referred to as Madelon, is 2000 rows and 500 features (1 index, 1 target). The dataset is artificial with a two-class target (-1, 1) with continuous input (parameter) variables.

### Problem Statement

Implement a machine learning pipeline designed to perform a naive logistic regression as a baseline model.

### Solution Statement

Provide a jupyter notebook that will control the baseline model pipeline with little (or no) regularization.

### Metric

The metric that I will use with my logistic regression is accuracy rate.

### Benchmark

The benchmark I will judge the baseline model against will be the baseline accuracy benchmark, which is the percentage of the database that would be guessed correctly if the majority target were applied to the entire dataset.

## Implementation

Implement the following code pipeline using the functions you write in `lib/project_5_ADH.py`.

<img src="assets/benchmarking.png" width="600px">

In [2]:
madelon_df = load_data_from_database('dsi_student', 'correct horse battery staple', 'joshuacook.me', 
                                     '5432', 'dsi', 'madelon')

In [3]:
madelon_df.head()

index  feat_000  feat_001  feat_002  feat_003  feat_004  feat_005  \
0      0       485       477       537       479       452       471   
1      1       483       458       460       487       587       475   
2      2       487       542       499       468       448       471   
3      3       480       491       510       485       495       472   
4      4       484       502       528       489       466       481   

   feat_006  feat_007  feat_008  ...    feat_491  feat_492  feat_493  \
0       491       476       475  ...         481       477       485   
1       526       479       485  ...         478       487       338   
2       442       478       480  ...         481       492       650   
3       417       474       502  ...         480       474       572   
4       402       478       487  ...         479       452       435   

   feat_494  feat_495  feat_496  feat_497  feat_498  feat_499  label  
0       511       485       481       479       475       496     -1  
1       513       486       483       492       510       517     -1  
2       506       501       480       489       499       498     -1  
3       454       469       475       482       494       461      1  
4       486       508       481       504       495       511      1  

[5 rows x 502 columns]

### Accuracy Score

In [4]:
madelon_df['label'].describe()

count    2000.00000
mean        0.00000
std         1.00025
min        -1.00000
25%        -1.00000
50%         0.00000
75%         1.00000
max         1.00000
Name: label, dtype: float64

### Modeling

In [5]:
data_dict = make_data_dict(madelon_df, random_state = 40, test_size = 0.20)

In [6]:
data_dict = general_transformer(StandardScaler(), data_dict)

In [7]:
lg_c_small = general_model(LogisticRegression(C = 1E10), data_dict)

In [8]:
data_dict['train score'], data_dict['test score']

(0.78125, 0.55249999999999999)

## Results

I beat our accuracy score of 50% with a test score of ~55%. 

Note, our accuracy score is 50% given the mean of the target ('label' column) is 0. Because the target is binary, either -1 or 1, we know that a perfect mean of 0.0 indicates that there is an equal amount of -1 and 1 in our target. 